contracts/hydro/src/contract.rs

```diff
@@ -2753,6 +2771,128 @@ pub fn convert_lockup_to_dtoken_reply(
         .add_attribute("issue_amount", issue_amount.to_string()))
 }
 
+/// Allows a user to buy out (i.e., pay off) the pending slash on a specific lockup.
+///
+/// This function:
+/// - Validates ownership of the lockup.
+/// - Loads the pending slash amount and calculates its value in base denomination. (ATOM)
+/// - Accepts user funds, converts them into base denomination using token group ratios,
+///   and determines if the slash is fully paid off.
+/// - If fully paid, removes the pending slash; otherwise, updates it with the remaining amount.
+/// - Tracks how much of each token was used and refunds any excess back to the user.
+pub fn buyout_pending_slash(
+    deps: DepsMut<NeutronQuery>,
+    env: Env,
+    info: MessageInfo,
+    constants: &Constants,
+    lock_id: u64,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    // Step 1: Get the lockup
+    let lockup = get_owned_lock_entry(deps.storage, &info.sender, lock_id)?;
+
+    // Step 2: Load pending slash
+    let Some(slash_amount) = LOCKS_PENDING_SLASHES.may_load(deps.storage, lock_id)? else {
+        return Err(new_generic_error(format!(
+            "Pending slash for lock id: {lock_id} doesn't exist.",
+        )));
+    };
+
+    let current_round = compute_current_round_id(&env, constants)?;
+    let mut token_manager = TokenManager::new(&deps.as_ref());
+
+    let slash_denom = lockup.funds.denom.clone();
+    let token_group_id = token_manager
+        .validate_denom(&deps.as_ref(), current_round, slash_denom.clone())
+        .map_err(|err| {
+            new_generic_error(format!("Error validating denom '{slash_denom}': {err}"))
+        })?;
+
+    let slash_ratio = token_manager
+        .get_token_group_ratio(&deps.as_ref(), current_round, token_group_id)
+        .map_err(|err| {
+            new_generic_error(format!(
+                "Error fetching token group ratio for '{slash_denom}': {err}",
+            ))
+        })?;
+
+    let slash_base_atom = Decimal::from_atomics(slash_amount.u128(), 18).unwrap() * slash_ratio;
```

Comment:
Let's use `Decimal::from_ratio(input, Uint128::one())` since this is used in majority of other places. Make this change throughout the entire function.

Also, use `checked_mul(slash_ratio)` istead of `* slash_ratio`